File: compiler/bench/prim.c
// expected: 499
#include <stdio.h>
#define N 500
#define INF 1000000

int adj[N][4];
int weight[N][4];
int deg[N];
int key[N];
char inMST[N];

void edge(int from, int to, int w) {
    adj[from][deg[from]] = to;
    weight[from][deg[from]] = w;
    deg[from]++;
}

void init_graph() {
    for (int i = 0; i < N; i++) deg[i] = 0;
    for (int i = 1; i < N; i++) {
        edge(i, i - 1, 1);
        edge(i - 1, i, 1);
    }
}

int find_min() {
    int minNode = -1, minKey = INF;
    for (int i = 0; i < N; i++) {
        if (!inMST[i] && key[i] < minKey) {
            minKey = key[i];
            minNode = i;
        }
    }
    return minNode;
}

int prim() {
    for (int i = 0; i < N; i++) {
        key[i] = INF;
        inMST[i] = 0;
    }
    key[0] = 0;
    int mst_weight = 0;
    for (int iter = 0; iter < N; iter++) {
        int u = find_min();
        if (u < 0) break;
        mst_weight += key[u];
        inMST[u] = 1;
        for (int i = 0; i < deg[u]; i++) {
            int v = adj[u][i];
            int w = weight[u][i];
            if (!inMST[v] && w < key[v]) {
                key[v] = w;
            }
        }
    }
    return mst_weight;
}

int main() {
    init_graph();
    printf("%d\n", prim());
    return 0;
}

File: compiler/bench/prim.fs
\ expected: 499
\ Prim MST, 500 nodes connected in line, MST weight = 499

500 constant N
1000000 constant INF
create adj N 4 * cells allot
create weight N 4 * cells allot
create deg N cells allot
create key N cells allot
create inMST N allot
variable mst-weight

: edge ( from to w -- )
  2 pick cells deg + @ 4 *
  3 pick + cells weight + !
  over cells deg + @ 4 *
  2 pick + cells adj + !
  swap cells deg + dup @ 1+ swap ! ;

: init-graph ( -- )
  N 0 do 0 i cells deg + ! loop
  N 1 do
    i i 1- 1 edge
    i 1- i 1 edge
  loop ;

: find-min ( -- node|-1 )
  -1 INF
  N 0 do
    i inMST + c@ 0= if
      i cells key + @ over < if
        2drop i i cells key + @
      then
    then
  loop
  drop ;

: prim ( -- )
  N 0 do
    INF i cells key + !
    0 i inMST + c!
  loop
  0 0 cells key + !
  0 mst-weight !
  N 0 do
    find-min dup 0< if drop unloop exit then
    dup cells key + @ mst-weight +!
    1 over inMST + c!
    dup cells deg + @ 0 ?do
      dup 4 * i + cells adj + @
      dup inMST + c@ 0= if
        2dup cells key + @
        2 pick 4 * i + cells weight + @
        over > if
          nip swap cells key + !
        else 2drop drop then
      else drop then
    loop
    drop
  loop ;

: main init-graph prim mst-weight @ . cr ;
